CMakeLists.txt

```diff
@@ -0,0 +1,178 @@
+cmake_minimum_required(VERSION 3.12)
+find_package(CUDA REQUIRED)
+
+include_directories(${PROJECT_SOURCE_DIR})
+set(CMAKE_LIBRARY_OUTPUT_DIRECTORY lib)
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY bin)
+set(CMAKE_BI)
+set(QUERY_MODE $ENV{QUERY_MODE})
+
+# set C_FLAGS and NVCC FLAGS
+set(CMAKE_C_FLAGS -fPIC)
+set(CMAKE_SHARED_LIBRARY_SUFFIX_C .so)
+set(CMAKE_SHARED_LIBRARY_SUFFIX_CXX .so)
+set(CUDA_NVCC_FLAGS "-lineinfo -I. -std=c++11 $ENV{NVCCFLAGS}")
+set(GENCODE_FLAGS "-gencode arch=compute_60,code=sm_60 -gencode arch=compute_60,code=compute_60")
+
+if (NOT QUERY_MODE)
+    execute_process(COMMAND which nvidia-smi && nvidia-smi | grep "Driver Version:" RESULT_VARIABLE ret)
+    if (NOT ${ret})
+        SET(QUERY_MODE DEVICE)
+    else ()
+        SET(QUERY_MODE HOST)
+    endif ()
+endif ()
+
+if (QUERY_MODE STREQUAL "DEVICE")
+    list(APPEND MARCROS "RUN_ON_DEVICE=1")
+    cuda_add_library(mem SHARED memutils/memory.h memutils/memory/cuda_malloc.cu)
+else ()
+    add_library(mem SHARED memutils/memory.h memutils/memory/malloc.c)
+endif ()
+
+# based on DEBUG and QUERY_MODE, transform CUDA_NVCC_FLAGS
+if (DEBUG STREQUAL "y")
+    set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS} -g -G --compiler-options='-g -ggdb'")
+else ()
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O3")
+endif ()
+
+string(REPLACE " " ";" CUDA_XCOMPILER_FLAGS ${CMAKE_C_FLAGS})
+list(TRANSFORM CUDA_XCOMPILER_FLAGS PREPEND " -Xcompiler ")
+string(REPLACE ";" " " CUDA_XCOMPILER_FLAGS ${CUDA_XCOMPILER_FLAGS})
+# transform macro to prepend -D prefix
+list(TRANSFORM MARCROS PREPEND "-D ")
+set(CUDA_NVCC_FLAGS "${CUDA_NVCC_FLAGS} ${CUDA_XCOMPILER_FLAGS} ${GENCODE_FLAGS} ${MARCROS}")
+file(MAKE_DIRECTORY ${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
+file(MAKE_DIRECTORY ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
+
+cuda_add_library(algorithm SHARED
+        query/algorithm.hpp
+        query/algorithm.cu
+        query/binder.hpp
+        query/dimension_transform.cu
+        query/filter.cu
+        query/functor.cu
+        query/geo_intersects.cu
+        query/hash_lookup.cu
+        query/hll.cu
+        query/iterator.hpp
+        query/measure_transform.cu
+        query/scratch_space_transform.cu
+        query/sort_reduce.cu
+        query/transform.cu
+        query/transform.hpp
+        query/utils.cu
+        query/utils.hpp)
+
+
+################################
+# GTest
+################################
+set(GTEST_ROOT $ENV{GTEST_ROOT})
+link_directories(${GTEST_ROOT})
```

Comment:
possible to merge to `ink_directories($ENV{GTEST_ROOT})` ?

Reply:
we reference GTEST_ROOT multiple times later, see target_include_directories(all_unittest PRIVATE ${GTEST_ROOT}/include ${gtest_SOURCE_DIR})